File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    if (intervals.length === 0) return intervals;

    const res = []
    intervals.sort((a,b) => a[0]-b[0])
    res.push(intervals[0])
    // [[1,4],[4,5]]
    // res = [[1,4]]
    for (let i = 1; i < intervals.length; i++) {
        // [4,5]
        const [start, end] = intervals[i]
        const prevEnd = res[res.length - 1][1]
        if (start <= prevEnd) {
            res[res.length - 1][1] = Math.max(end, prevEnd)
        } else {
            res.push(intervals[i])
        }
    }
    return res
};

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()

        res = []
        prev = intervals[0]
        res.append(prev)

        for i in range(1, len(intervals)):
            interval = intervals[i]
            start= interval[0]
            end = interval[1]
            if start <= prev[1]:
                prev[1] = max(end, prev[1])
            else:
                prev = interval
                res.append(interval)
        
        return res



# var merge = function(intervals) {
#     intervals.sort((a,b) => a[0] - b[0])
#     let res = []
#     let prev = intervals[0]
#     res[0] = prev

#     for (let i = 1; i < intervals.length;i++) {
#         const interval = intervals[i]
#         const [start,end] = interval
#         if (start <= prev[1]) {
#             prev[1] = Math.max(end, prev[1])
#         } else {
#             prev = interval
#             res.push(prev)
#         }
#     }
#     return res
# };